File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        // Dummy head simplifies edge cases; `last` always points to last node of merged list.
        ListNode dummy(0);
        ListNode* last = &dummy;

        // Invariant: dummy.next..last is the merged sorted prefix of nodes consumed from list1/list2.
        while (list1 && list2) {
            if (list1->val <= list2->val) {
                last->next = list1;      // splice node from list1
                list1 = list1->next;     // advance list1
            } else {
                last->next = list2;      // splice node from list2
                list2 = list2->next;     // advance list2
            }
            last = last->next;           // advance tail
        }

        // Append the remaining nodes (already sorted).
        last->next = list1 ? list1 : list2;

        return dummy.next;
    }
};

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        dummy = ListNode()
        last  = dummy
        while list1 and list2:
            if list1.val <= list2.val:
                last  = list1
                list1 = list1.next
            else:
                last  = list2
                list2 = list2.next
            last = last.next
        return dummy.next